framework/src/util/SharedLibrary.cpp

```diff
@@ -99,25 +101,31 @@ void SharedLibrary::Load(int flags)
 #ifdef US_PLATFORM_POSIX
   d->m_Handle = dlopen(libPath.c_str(), flags);
   if (!d->m_Handle) {
+    std::error_code err_code(errno, std::generic_category());
     std::string err_msg = "Error loading " + libPath + ".";
     const char* err = dlerror();
     if (err) {
       err_msg += " " + std::string(err);
     }
-
-    throw std::runtime_error(err_msg);
+    
+    // Bundle of origin information is not available here. It will be
+    // set later in BundlePrivate::Start0() in a catch statement.
+    throw cppmicroservices::SharedLibraryException(err_code, err_msg);
   }
 #else
   US_UNUSED(flags);
   std::wstring wpath(cppmicroservices::util::ToWString(libPath));
   d->m_Handle = LoadLibraryW(wpath.c_str());
   if (!d->m_Handle) {
+    std::error_code err_code(GetLastError(), std::generic_category());
     std::string errMsg = "Loading ";
     errMsg.append(libPath)
       .append("failed with error: ")
       .append(util::GetLastWin32ErrorStr());
-
-    throw std::runtime_error(errMsg);
+    
+    // Bundle of origin information is not available here. It will be
+    // set later in BundlePrivate::Start0() in a catch statement.
+    throw cppmicroservices::SharedLibraryException(err_code, errMsg);
```

Comment:
catching the `std::runtime_error` in the caller and then re-throwing a `SharedLibraryException` can remove the need to have `SharedLibraryException::SetBundle`

Reply:
I think it might be a better idea to throw a std::system_error, then rethrow as a SharedLibraryException. Actually that's how I had it implemented first, then thought about implications of this. This would mean that we assume all the code in BundlePrivate::Start0() would be expected to NOT throw a system_error except when calling SharedLibrary::Load(int flags).

I'll make the changes accordingly with std::system_error.